src/cancelq/posix/cancellable.c

```diff
@@ -0,0 +1,399 @@
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <pthread.h>
+
+#include "cancellable.h"
+
+#ifndef HEAP_CHECK
+#define HEAP_CHECK true // Set to false to save oodles of execution time.
+#endif
+
+#define BASE_HEAP_SIZE 25
+#define INITIAL_ID_COUNT 16
+
+#define UNUSED __attribute__((unused))
+
+// Normal events that can't be cancelled
+typedef struct
+{
+    const void *event;
+} event_t;
+
+// Events that can be cancelled, which includes events whose sending
+// can be delayed
+typedef struct
+{
+    event_t event;
+    cancellable_id_t id;
+    abs_time_t delivery_time;
+} cancellable_t;
+
+typedef cancellable_t heap_data_t;
+#include "heap.c" // Note the .c
+
+typedef enum
+{
+    ID_UNUSED = 0,
+    ID_DELIVERED,
+    ID_WAITING
+} id_state_t;
+
+struct event_queue_s
+{
+    heap_t heap;
+    // Current size of the array of IDs
+    cancellable_id_t idsize;
+
+    // Array of IDs for cancellation. The ID is an index into this array.
+    id_state_t *ids;
+
+    pthread_mutex_t lock;
+    pthread_cond_t empty;
+};
+
+UNUSED static void print_queue(event_queue_t *q)
+{
+    print_heap(&q->heap);
+}
+
+static bool empty_LH(const event_queue_t *q)
+{
+    return heap_empty(&q->heap);
+}
+
+static cancellable_id_t new_id(event_queue_t *q)
+{
+    id_state_t *new_ids;
+    cancellable_id_t new_size;
+    cancellable_id_t id;
+
+    for (id = 0; id < q->idsize; id++)
+    {
+        if (q->ids[id] == ID_UNUSED)
+        {
+            q->ids[id] = ID_WAITING;
+            return id;
+        }
+    }
+
+    // We're out of IDs. Double the size of the array for now.
+    new_size = q->idsize * 2;
+    new_ids = realloc(q->ids, sizeof(id_state_t) * new_size);
+    if (new_ids == NULL)
+        return NOT_CANCELLABLE;
```

Comment:
This use of `NOT_CANCELLABLE` gives that id two distinct connotations.  Is there a particular reason you chose it over `SCHEDULE_FAIL` or some third special error code?

Reply:
At first, I had `eq_schedule` returning `NOT_CANCELLABLE` if the ID acquisition failed, but still scheduling it. Then I thought about why anyone would be ok with having their event scheduled even though they can't cancel it and changed it.

I left it this way because it's the right ID to return if an event can't be cancelled. Since `new_id` is static, I didn't think it was important enough to justify another error code.